plugin/pkg/auth/authenticator/token/openid/openid.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// openid implements the authenticator.Token interface using the OpenID protocol.
+package openid
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/coreos/go-oidc/jose"
+	"github.com/coreos/go-oidc/oidc"
+	"github.com/golang/glog"
+)
+
+var (
+	maxRetries   = 5
+	retryBackoff = time.Second * 3
+)
+
+var errNoKeyEndpoint = fmt.Errorf("OpenID provider must provide 'jwks_uri' for public key discovery")
+
+type OpenIDAuthenticator struct {
+	client *oidc.Client
+}
+
+// NewOpenID creates a new OpenID client with the given clientID and discovery URL.
+// NOTE(yifan): For now we assume the server provides the "jwks_uri" so we don't
+// need to manager the key sets by ourselves.
+func NewOpenID(clientID, discovery, caFile string) (*OpenIDAuthenticator, error) {
+	var cfg oidc.ProviderConfig
+	var err error
+	var roots *x509.CertPool
+
+	url, err := url.Parse(discovery)
+	if err != nil {
+		return nil, err
+	}
+
+	if url.Scheme == "http" {
+		glog.Warning("Using insecure connection to the OpenID service")
```

Comment:
Are we sure that we want to allow this? If we don't have a good use case for it, we should just not support it as it is clearly a security issue.